File: runtime/py/transport/HTTPClientTransport.py
import http.client
import io
import urllib.request
import urllib.error

from myrpc.Common import MessageTruncatedException
from myrpc.transport.TransportBase import TransportState, TransportBase, TransportException

class HTTPClientTransport(TransportBase):
    """Provide HTTP client transport."""

    def __init__(self, url):
        super().__init__()

        self._url = url
        self._timeout = None
        self._opener = None

    def set_state(self, state):
        if state == TransportState.WRITE_BEGIN:
            self._reset()
        elif state == TransportState.WRITE_END:
            self._flush()

    def read(self, count):
        buf = self._rf.read(count)
        if len(buf) < count:
            raise MessageTruncatedException()

        return buf

    def write(self, buf):
        self._wf.write(buf)

    def set_opener(self, opener):
        self._opener = opener

    def set_timeout(self, timeout):
        self._timeout = timeout

    def _reset(self):
        self._rf = None
        self._wf = io.BytesIO()

    def _flush(self):
        wbuf = self._wf.getvalue()

        req = urllib.request.Request(self._url, data = wbuf, method = "POST")
        req.add_header("Content-Type", "application/octet-stream")

        opener = self._opener.open if self._opener else urllib.request.urlopen

        try:
            resp = opener(req, timeout = self._timeout)
            rbuf = resp.read()
        except (urllib.error.URLError, http.client.HTTPException, OSError) as e:
            # FIXME: Is it good idea to convert all of those exceptions to
            # HTTPClientException?

            raise HTTPClientException(e)

        self._rf = io.BytesIO(rbuf)

class HTTPClientException(TransportException):
    """Exception class for HTTP-related errors."""

    def __init__(self, reason):
        super().__init__(str(reason))

        self._reason = reason

    def get_reason(self):
        return self._reason

File: runtime/js/transport/HTTPClientTransport.js
// FIXME: HTTPClientTransport is not yet supported under Node.js.

myrpc.transport.HTTPClientTransport = function(url)
{
    myrpc.transport.TransportBase.call(this);

    this._url = url;
    this._xhr = null;
    this._oncontinue = null;
    this._timeout = null;
};

myrpc.transport.HTTPClientTransport.prototype = Object.create(myrpc.transport.TransportBase.prototype);
myrpc.transport.HTTPClientTransport.prototype.constructor = myrpc.transport.HTTPClientTransport;

myrpc.transport.HTTPClientTransport.prototype.set_state = function(state)
{
    switch (state) {
    case myrpc.transport.TransportState.WRITE_BEGIN:
	this._reset();
	break;

    case myrpc.transport.TransportState.WRITE_END:
	this._flush();
	break;

    case myrpc.transport.TransportState.READ_BEGIN:
	this._check_status();
	break;
    }
};

myrpc.transport.HTTPClientTransport.prototype.read = function(count)
{
    var buf = this._rbio.read(count);

    if (count > buf.length)
	throw new myrpc.common.MessageTruncatedException();

    return buf;
};

myrpc.transport.HTTPClientTransport.prototype.write = function(buf)
{
    this._wbio.write(buf);
};

myrpc.transport.HTTPClientTransport.prototype.set_oncontinue = function(oncontinue)
{
    this._oncontinue = oncontinue;
};

myrpc.transport.HTTPClientTransport.prototype.set_timeout = function(timeout)
{
    this._timeout = timeout;
};

myrpc.transport.HTTPClientTransport.prototype._reset = function()
{
    // Abort running request and reset transport.

    if (this._xhr) {
	this._xhr.abort();
	this._xhr = null;
    }

    this._rbio = null;
    this._wbio = new myrpc.util.BufferIO();
};

myrpc.transport.HTTPClientTransport.prototype._flush = function()
{
    var wbuf = this._wbio.get_buffer();
    var that = this;

    if (!this._oncontinue)
	throw new myrpc.common.MyRPCInternalException("oncontinue callback is null");

    this._xhr = new XMLHttpRequest();
    this._xhr.open("POST", this._url, true);
    this._xhr.setRequestHeader("Content-Type", "application/octet-stream");
    this._xhr.responseType = "arraybuffer";
    this._xhr.onreadystatechange = function() {
	var rbuf;

	if (that._xhr.readyState == 4) {
	    that._status = that._xhr.status;
	    if (that._is_status_ok()) {
		rbuf = new Uint8Array(that._xhr.response);
		that._rbio = new myrpc.util.BufferIO(rbuf);
	    }

	    // Do not abort already completed request (see _reset).
	    that._xhr = null;

	    that._oncontinue();
	}
    };
    if (this._timeout != null)
	this._xhr.timeout = this._timeout;
    this._xhr.send(wbuf);
};

myrpc.transport.HTTPClientTransport.prototype._check_status = function()
{
    if (!this._is_status_ok())
	throw new myrpc.transport.TransportException("HTTP request failed with status code " + this._status);
};

myrpc.transport.HTTPClientTransport.prototype._is_status_ok = function()
{
    var r = (this._status >= 200 && this._status < 300);

    return r;
};
